cpp/ql/src/semmle/code/cpp/dataflow/internal/FlowVar.qll

```diff
@@ -243,6 +244,54 @@ private module PartialDefinitions {
     }
   }
 
+  class SmartPointerPartialDefinition extends PartialDefinition {
+    Variable pointer;
+    Expr innerDefinedExpr;
+
+    SmartPointerPartialDefinition() {
+      exists(Expr convertedInner |
+        not this instanceof Conversion and
+        valueToUpdate(convertedInner, this.getFullyConverted(), node) and
+        innerDefinedExpr = convertedInner.getUnconverted() and
+        innerDefinedExpr = getAPointerWrapperAccess(pointer)
+      )
```

Comment:
Is this case fully subsumed by the charpred of `VariablePartialDefinition`?

Reply:
It is. This means that `*smart_ptr = e` is already a partial definition according to `VariablePartialDefinition`. The problem is that for `VariablePartialDefinition` it's a partial definition _of the variable_. Whereas for `SmartPointerPartialDefinition` it's a partial definition _of the call to `operator*`_. The same problem arises in `IteratorPartialDefinition`. I'm not exactly sure how to fix this short of making `valueToUpdate` more complex. To reduce duplication I've refactored the common parts out into a predicate in 0a6aef71a2f957386f895d594f227a0c67c387be.